hikari/impl/event_manager_base.py

```diff
@@ -233,37 +259,137 @@ def _assert_is_listener(parameters: typing.Iterator[inspect.Parameter], /) -> No
     if next(parameters, None) is None:
         raise TypeError("Event listener must have one positional argument for the event object.")
 
-    if any(param.default is not inspect.Parameter.empty for param in parameters):
+    if any(param.default is inspect.Parameter.empty for param in parameters):
         raise TypeError("Only the first argument for a listener can be required, the event argument.")
 
 
+def filtered(
+    event_types: typing.Union[typing.Type[base_events.Event], typing.Sequence[typing.Type[base_events.Event]]],
+    cache_components: config.CacheComponents = config.CacheComponents.NONE,
+    /,
+) -> typing.Callable[[UnboundMethodT[EventManagerBaseT]], UnboundMethodT[EventManagerBaseT]]:
+    """Add metadata to a consumer method to indicate when it should be unmarshalled and dispatched.
+
+    Parameters
+    ----------
+    event_types
+        Types of the events this raw consumer method may dispatch.
+        This may either be a singular type of a sequence of types.
+
+    Other Parameters
+    ----------------
+    cache_components : hikari.config.CacheComponents
+        Bitfield of the cache components this event may make altering calls to.
+        This defaults to `hikari.config.CacheComponents.NONE`.
+    """
+    if isinstance(event_types, typing.Sequence):
+        # dict.fromkeys is used to remove any duplicate entries here
+        event_types = tuple(dict.fromkeys(itertools.chain.from_iterable(e.dispatches() for e in event_types)))
```

Comment:
Using a set instead of a tuple here can help with base events, as they dispatch a lot of other events

(Same applies for `.dispatches()`)

Reply:
That'd erm use noticeable more memory and be slower to iterate over which is the main thing we do with it